### PR TITLE
refactor block log

### PR DESCRIPTION
**Pull request by huangminghuang**
## Change Description
The PR creates some new abstractions to facilitate code use for parsing the block log/index files.

- block_log_preamble: used to serialize/deserialize the first section of the blocks.log file
- block_log_data: the read only view of the blocks.log file
- block_log_index: the read only view of the blocks.index file
- block_log_archive: the collection of block_log_data and block_log_index with basic check for the consistency between the two
- reverse_block_position_iterator: Used to traverse the block position (i.e. the last 8 bytes in each block log entry) of the blocks.log file

Most of the functionalities like trim/repair/smoke tests are built on top the new abstractions

The block_log_impl is also refactored so that  block_file.is_open() && index_file.is_open() is the invariant of the class. This eliminates the
repetitive checking whether the files are open in each method.

unittest code refactor

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: libraries/chain/block_log.cpp [hunk 1]
```diff
@@ -120,221 +163,310 @@ namespace eosio { namespace chain {
...
+         if (first_block_pos == file.size())
+            return 0;
+         return last_block_num() - first_block_num() + 1;
+      }
+
+      /**
+       *  Validate a block log entry WITHOUT deserializing the entire block data. 
+       **/
+      void light_validate_block_entry_at(uint64_t pos, uint32_t expected_block_num) const {
+         uint32_t actual_block_num = block_num_at(pos);
```
Comment (brianjohnson5972) [inline libraries/chain/block_log.cpp:314]: make const

File: libraries/chain/block_log.cpp [hunk 2]
```diff
@@ -1155,320 +872,112 @@ namespace eosio { namespace chain {
...
-         interval = std::max((td.last_block + 8 - td.first_block) >> 3, 1U);
+         interval = std::max((archive.log_index.num_blocks() + 7) >> 3, 1);
       }
-      for (uint32_t n = td.first_block; ; n += interval) {
-         if (n > td.last_block)
-            n = td.last_block;
-         td.block_pos(n);                                 //check block 'n' is where blocks.index says
-         if (n == td.last_block)
-            break;
+      uint32_t expected_block_num = archive.log_data.first_block_num();
```
Comment (brianjohnson5972) [inline libraries/chain/block_log.cpp:976]: make const

Comment (huangminghuang) [reply libraries/chain/block_log.cpp:991]: It's modified, shouldn't be const.

Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp:991]: whoops

**Pull request merged**